crates/accelerate/src/rayon_ext.rs

```diff
@@ -0,0 +1,161 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2023
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+//! Extension structs for use with Rayon parallelism.
+
+use rayon::iter::plumbing::*;
+use rayon::prelude::*;
+
+pub trait ParallelSliceMutExt<T: Send>: ParallelSliceMut<T> {
+    /// Create a parallel iterator over mutable chunks of uneven lengths for this iterator.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the sums of the given lengths do not add up to the length of the slice.
+    fn par_uneven_chunks_mut<'len, 'data>(
+        &'data mut self,
+        chunk_lengths: &'len [usize],
+    ) -> ParUnevenChunksMut<'len, 'data, T> {
+        let mut_slice = self.as_parallel_slice_mut();
+        let chunk_sum = chunk_lengths.iter().sum::<usize>();
+        let slice_len = mut_slice.len();
+        if chunk_sum != slice_len {
+            panic!("given slices of total size {chunk_sum} for a chunk of length {slice_len}");
+        }
+        ParUnevenChunksMut {
+            chunk_lengths,
+            data: mut_slice,
+        }
+    }
+}
+
+impl<T: Send, S: ?Sized> ParallelSliceMutExt<T> for S where S: ParallelSliceMut<T> {}
```

Comment:
Why did I write `<S: ?Sized> ... where S: ParallelSliceMut<T>` and not put both either in the `<>` or in the `where`?  Who knows.